File: 0097 - Interleaving String/solution.c
bool isInterleave(char * s1, char * s2, char * s3){
    int i, j, dp[101] = {};
    for(i = 0; !i || s1[i-1]; i++){
        for(j = 0; !j || s2[j-1]; j++){
            if(i+j && !s3[i+j-1]) return false;
            if(i==0 && j==0) dp[j] = 1;
            else if(i == 0) dp[j] = dp[j-1]&&s2[j-1]==s3[j-1];
            else if(j == 0) dp[j] = dp[j]&&s1[i-1]==s3[i-1];
            else dp[j] = (dp[j-1]&&s2[j-1]==s3[i+j-1])||(dp[j]&&s1[i-1]==s3[i+j-1]);
        }
    }
    return !s3[i+j-2] && dp[j-1];
}

File: 0097 - Interleaving String/solution.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        @lru_cache(None)
        def dp(i, j):
            if i == 0 and j == 0:
                return True
            if i == 0:
                return s2[j-1] == s3[j-1] and dp(i, j-1)
            if j == 0:
                return s1[i-1] == s3[i-1] and dp(i-1, j)
            return s2[j-1] == s3[i+j-1] and dp(i, j-1) or s1[i-1] == s3[i+j-1] and dp(i-1, j)
        n1, n2, n3 = len(s1), len(s2), len(s3)
        return n1+n2 == n3 and dp(n1, n2)